peptide_normalization.py

```diff
@@ -111,72 +124,96 @@ def intensity_normalization(dataset: DataFrame, field: str, class_field: str = "
     else:
-        # normalize taking into account samples
-        normalize_df = dataset[[PEPTIDE_SEQUENCE, CONDITION, field, class_field]]
-        # group peptide + charge into a single peptide intensity using the mean.
-        normalize_df = pd.pivot_table(normalize_df, values=field, index=[PEPTIDE_SEQUENCE, CONDITION],
-                                      columns=class_field,
-                                      aggfunc={field: np.mean})
-
-        print(normalize_df.head())
-
-        # Remove all peptides in less than 30% of the samples.
-        if remove_peptides:
-            n_samples = len(normalize_df.columns)
-            normalize_df = normalize_df.dropna(thresh=round(n_samples * 0.3))
-
-        # Imputation of the values using KNNImputer
-        # (https://scikit-learn.org/0.16/modules/generated/sklearn.preprocessing.Imputer.html)
-        if imputation:
-            imputer = SimpleImputer()
-            if imputation_method != "simple":
-                imputer = KNNImputer(n_neighbors=2, weights="uniform")
-            normalized_matrix = imputer.fit_transform(normalize_df)
-            if scaling_method != "quantile":
-                normalized_matrix = scaler.fit_transform(normalized_matrix)
-            else:
-                normalized_matrix = qnorm.quantile_normalize(normalized_matrix)
-        else:
-            normalized_matrix = scaler.fit_transform(normalize_df)
-
-        normalize_df[:] = normalized_matrix
-        normalize_df = normalize_df.reset_index()
-        normalize_df = normalize_df.melt(id_vars=[PEPTIDE_SEQUENCE, CONDITION])
-        normalize_df.rename(columns={'value': NORM_INTENSITY}, inplace=True)
-        dataset = pd.merge(dataset, normalize_df, how='left', on=[PEPTIDE_SEQUENCE, CONDITION, class_field])
-        dataset.pop(NORM_INTENSITY + "_x")
-        dataset.rename(
-            columns={NORM_INTENSITY + "_y": NORM_INTENSITY}, inplace=True)
+        dataset = dataset.loc[dataset.groupby([PEPTIDE_SEQUENCE, PEPTIDE_CHARGE, SAMPLE_ID, BIOREPLICATE])[
+            SEARCH_ENGINE].idxmax()].reset_index(drop=True)
+    return dataset
+
+def sum_peptidoform_intensities(dataset: DataFrame) -> DataFrame:
+    """
+    Sum the peptidoform intensities for all peptidofrom across replicates of the same sample.
 
+    :param dataset: Dataframe to be analyzed
+    :return: dataframe with the intensities
+    """
+    dataset = dataset[dataset[NORM_INTENSITY].notna()]
+    dataset = dataset.groupby([PEPTIDE_SEQUENCE, PEPTIDE_CHARGE, SAMPLE_ID])[NORM_INTENSITY].sum()
+    dataset = dataset.reset_index()
     return dataset
 
+def average_peptide_intensities(dataset: DataFrame) -> DataFrame:
+    """
+    Median the intensities of all the peptidoforms for an specific peptide sample combination.
+    :param dataset: Dataframe containing all the peptidoforms
+    :return: New dataframe
+    """
+    dataset = dataset.groupby([PEPTIDE_CANONICAL, SAMPLE_ID])[NORM_INTENSITY].median()
```

Comment:
This should group by canonical and sample

Reply:
It is grouping by Canonical and SampleID no?

Reply:
ah yes. I had written something else first and corrected myself.